src/openMVG/voctree/database.cpp

```diff
@@ -169,83 +172,258 @@ void Database::loadWeights(const std::string& file)
   }
 }
 
-/**
- * Given a list of visual words associated to the features of a document it computes the 
- * vector of unique weighted visual words
- * 
- * @param[in] document a list of (possibly repeated) visual words
- * @param[out] v the vector of visual words (ie the visual word histogram of the image)
- */
-void Database::computeVector(const std::vector<Word>& document, DocumentVector& v) const
-{
-  //for each visual word in the list
-  for(const Word &word : document)
-  {
-    // update its weighted count inside the map
-    // the map v contains only the visual words that are associated to some features
-    // the visual words in v are unique unlikely the document
-    if(v.find(word) == v.end())
-      v[word] = word_weights_[word];
-    else
-      v[word] += word_weights_[word];
-  }
-  normalize(v);
-}
+///**
+// * Normalize a document vector representing the histogram of visual words for a given image
+// * 
+// * @param[in/out] v the unnormalized histogram of visual words
+// */
+//void Database::normalize(SparseHistogram& v) const
+//{
+//  float sum = 0.0f;
+//  for(SparseHistogram::iterator i = v.begin(), ie = v.end(); i != ie; ++i)
+//    sum += i->second;
+//  float inv_sum = 1.0f / sum;
+//  for(SparseHistogram::iterator i = v.begin(), ie = v.end(); i != ie; ++i)
+//    i->second *= inv_sum;
+//}
 
 /**
- * Normalize a document vector representing the histogram of visual words for a given image
+ * Compute the distance between two histograms
  * 
- * @param[in/out] v the unnormalized histogram of visual words
+ * @param[in] v1 the first histogram
+ * @param[in] v2 the second histogram
+ * @param[in] distanceMethod the method used to compute distance between histograms
  */
-void Database::normalize(DocumentVector& v)
+float Database::sparseDistance(const SparseHistogram& v1, const SparseHistogram& v2, const std::string &distanceMethod) const
 {
-  float sum = 0.0f;
-  for(DocumentVector::iterator i = v.begin(), ie = v.end(); i != ie; ++i)
-    sum += i->second;
-  float inv_sum = 1.0f / sum;
-  for(DocumentVector::iterator i = v.begin(), ie = v.end(); i != ie; ++i)
-    i->second *= inv_sum;
-}
 
-float Database::sparseDistance(const DocumentVector& v1, const DocumentVector& v2)
-{
   float distance = 0.0f;
-  DocumentVector::const_iterator i1 = v1.begin(), i1e = v1.end();
-  DocumentVector::const_iterator i2 = v2.begin(), i2e = v2.end();
+  float epsilon = 0.001;
+  
+  SparseHistogram::const_iterator i1 = v1.begin(), i1e = v1.end();
+  SparseHistogram::const_iterator i2 = v2.begin(), i2e = v2.end();
+  
+  if(distanceMethod.compare("classic") == 0) 
```

Comment:
`distanceMethod == "classic"`  (seems easier to me, although they are equivalent)
Also, but this requires more refactoring, we can pass an object that actually compute the distance instead of a string (or even make it a template), i don't know something like

```c++
// the "classic" one
struct HistogramDistance{
  typedef SparseHistogram::const_iterator iter;   // jsut because i'm lazy
  static float distance(const iter &i1, const iter &i1e, const iter &i2, const iter &i2e )
  {
     // the code here
  }

};

// and then define all the other as derived class of HistogramDistance

// sparseDistance becomes
float Database::sparseDistance(..., HistogramDistance *distance){...}
// or
template<class distance=HistogramDistance >
float Database::sparseDistance(...);
```